### PR TITLE
EVP_DigestFinalXOF must not reset the EVP_MD_CTX

### PR DESCRIPTION
**Pull request by t8m**
It does not do it in legacy path and 1.1.1 so that must not change.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by beldmit — APPROVED** [review]
LGTM if tests are passed

**Comment by kroeckx** [comment]
Maybe this needs a test?

File: test/evp_extra_test.c
Comment (slontis) [inline test/evp_extra_test.c:691]: Maybe either add a comment or change the name of the test so that we know what this test was added for in the future..

File: crypto/evp/digest.c
Comment (slontis) [inline crypto/evp/digest.c:176]: == NULL? (Yes I know you just moved it :))

Comment (t8m) [reply crypto/evp/digest.c]: Fixed

File: test/evp_extra_test.c
Comment (t8m) [reply test/evp_extra_test.c:694]: Comment added

**Review by beldmit — APPROVED** [review]
The code looks good to me. I approve if tests passed.
I don't like the logic of reuse context after finalization, but it's out of the scope of this PR and a separate matter in total.

**Comment by t8m** [comment]
I do not see any problems with context reuse as long as it is not the incorrect reuse without proper DigestInit(_ex) call. It avoids re-doing some operations that would need to be done on reset context.

File: doc/man3/EVP_DigestInit.pod
Comment (slontis) [inline doc/man3/EVP_DigestInit.pod:211]: Maybe scramble around a bit?? I will leave it up to you..

I<type> can be NULL if I<ctx> has already been initialized via a previous call to EVP_DigestInit_ex()

File: doc/man3/EVP_DigestInit.pod
Comment (slontis) [inline doc/man3/EVP_DigestInit.pod:247]: I<type> of NULL?

File: test/evp_extra_test.c
Comment (slontis) [inline test/evp_extra_test.c:715]: small nit /* should be on a line by itself

**Comment by slontis** [comment]
Travis error does not look related to this PR.

**Comment by t8m** [comment]
@slontis comments addressed, do you approve?

@beldmit do you still approve?

**Comment by beldmit** [comment]
Yes, my approval still persists

**Comment by kroeckx** [comment]
The test I was expecting is that we actually call EVP_DigestFinalXOF() twice, to see that it produces the correct output.

**Comment by kroeckx** [comment]
> The test I was expecting is that we actually call EVP_DigestFinalXOF() twice, to see that it produces the correct output.

It seems that's not going to work, so forget about that part.

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
I did a clean rebase to just run the CI

**Comment by slontis** [comment]
Still approved.

**Comment by t8m** [comment]
CI failures unrelated

**Comment by t8m** [comment]
Merged to master

**Pull request closed without merging**
